### PR TITLE
refactor(source): all in async-stream

### PR DESCRIPTION
**Pull request by wangrunji0408**
I hereby agree to the terms of the [Singularity Data, Inc. Contributor License Agreement](https://gist.github.com/skyzh/0663682a70b0edde7ae991492f2314cb#file-s9y_cla).

## What's changed and what's your intention?

This PR changes all source readers to async-stream, in order to be more consistent with executors and prevent omitting events caused by select.

~~This also includes a significant change that the stream now yields **ONE** `SourceMessage` at a time instead of a batch of them. The point of batching is now delayed to after merging splits. This will make back pressure more accurate and avoid memory allocation of Vec. But I'm not sure whether it has any impact. Feel free to comment below about it.~~
(reverted due to performance issue, leave it for future PR)

## Checklist

- [x] I have written necessary rustdoc comments
- [x] I have added necessary unit tests and integration tests
- [x] All checks passed in `./risedev check` (or alias, `./risedev c`)

### CODE REVIEW

**Comment by TennyZhuang** [comment]
The PR is too excited..

File: src/source/src/connector_source.rs [hunk 1]
```diff
@@ -261,19 +216,12 @@ impl ConnectorSource {
...
             .await?;
 
-        let streams = readers
-            .into_iter()
-            .map(inner_connector_source_reader_into_stream)
-            .collect::<Vec<_>>()
-            .merge()
-            .into_stream()
-            .boxed();
+        let stream = select_all(readers.into_iter().map(|r| r.into_stream())).boxed();
```
Comment (BugenZhao) [inline src/source/src/connector_source.rs:219]: Compared to `select_all`, `merge` will shuffle the streams before each polling. Not sure whether this matters.

File: src/source/src/connector_source.rs [hunk 2]
```diff
@@ -113,93 +112,49 @@ impl InnerConnectorSourceReader {
...
-        for msg in batch {
-            if let Some(content) = msg.payload {
-                split_offset_mapping.insert(msg.split_id, msg.offset);
-                let writer = builder.row_writer();
-                match self.parser.parse(content.as_ref(), writer) {
-                    Err(e) => {
+    #[try_stream(boxed, ok = StreamChunkWithState, error = RwError)]
+    pub async fn into_stream(self) {
+        #[for_await]
+        for batch in self.stream.ready_chunks(1000) {
```
Comment (BugenZhao) [inline src/source/src/connector_source.rs:138]: May extract this as a constant, or make it configurable?

**Review by BugenZhao — COMMENTED** [review]
Generally LGTM 🚀🚀🚀

File: src/connector/src/source/base.rs [hunk 3]
```diff
@@ -76,9 +77,11 @@ pub trait SplitReader: Sized {
         columns: Option<Vec<Column>>,
     ) -> Result<Self>;
 
-    async fn next(&mut self) -> Result<Option<Vec<SourceMessage>>>;
+    fn into_stream(self) -> BoxStream<'static, Result<SourceMessage>>;
 }
 
+pub type BoxSourceStream = BoxStream<'static, Result<SourceMessage>>;
```
[the remark this answers is not in the document]
Comment (wangrunji0408) [reply src/connector/src/source/base.rs:83]: Seems like a serious problem. 😢 +1 with the proposal.

File: src/source/src/connector_source.rs (hunk 1 above)
Comment (wangrunji0408) [reply src/source/src/connector_source.rs:219]: ... which means that `select_all` has a potential starvation problem?

File: src/connector/src/source/base.rs (hunk 3 above)
[the remark this answers is not in the document]
Comment (BugenZhao) [reply src/connector/src/source/base.rs:83]: How about making the per-message stream a static one, but add a dynamic utility stream for batching them?

File: src/source/src/connector_source.rs (hunk 1 above)
Comment (BugenZhao) [reply src/source/src/connector_source.rs:219]: I think so. 🥵

Comment (xxchan) [reply src/source/src/connector_source.rs:221]: `risingwave_common::util::select_all` 🤔

Comment (BugenZhao) [reply src/source/src/connector_source.rs:221]: Cool. 😄

Comment (wangrunji0408) [reply src/source/src/connector_source.rs:222]: replaced by it

Comment (xxchan) [reply src/source/src/connector_source.rs:222]: Will we always prefer a fair `select_all`? :think:

**Comment by xxchan** [comment]
Suggested edit:
```diff
src/connector/src/source/kafka/source/reader.rs --- 1/1 --- Rust
41         state: ConnectorState,
42         _columns: Option<Vec<Column>>,
43     ) -> Result<Self> {
44         let bootstrap_servers = properties.brokers;  // suggest edit
45 
46         let mut config = ClientConfig::new();
47 


```

**Comment by xxchan** [comment]
Suggested edit:
```diff
src/connector/src/source/kafka/source/reader.rs --- 1/1 --- Rust
41 41         state: ConnectorState,
42 42         _columns: Option<Vec<Column>>,
43 43     ) -> Result<Self> {
[2m.. [0m44         // suggest edit
44 45         let bootstrap_servers = properties.brokers; // aa
45 46 
46 47         let mut config = ClientConfig::new();
47 48 


```

**Comment by xxchan** [comment]
Suggested edit:
```diff
src/connector/src/source/kafka/source/reader.rs --- 1/1 --- Rust
41 41         state: ConnectorState,
42 42         _columns: Option<Vec<Column>>,
43 43     ) -> Result<Self> {
[2m.. [0m44         // suggest edit
44 45         let bootstrap_servers = properties.brokers; // aa
45 46 
46 47         let mut config = ClientConfig::new();
47 48 


```

**Comment by xxchan** [comment]
Suggested edit:
```diff
src/connector/src/source/kafka/source/reader.rs --- 1/1 --- Rust
41 41         state: ConnectorState,
42 42         _columns: Option<Vec<Column>>,
43 43     ) -> Result<Self> {
[2m.. [0m44         // suggest edit
44 45         let bootstrap_servers = properties.brokers; // aa
45 46 
46 47         let mut config = ClientConfig::new();
47 48 


```

File: src/source/src/connector_source.rs (hunk 1 above)
Comment (BugenZhao) [reply src/source/src/connector_source.rs:222]: I think so. We tend to `select` on **unbounded** streams, where fairness should be critical.

Comment (xxchan) [reply src/source/src/connector_source.rs:221]: Then we can add a `disallowed-method` 😄

Comment (wangrunji0408) [reply src/source/src/connector_source.rs:221]: I've added it in 07ee20a. But it didn't give any warning when I run `./risedev c`. Any problems here?

Comment (xxchan) [reply src/source/src/connector_source.rs:221]: Did you `cargo clean`?

Comment (wangrunji0408) [reply src/source/src/connector_source.rs:221]: Nope. Why a clean is required? 😇

Comment (BugenZhao) [reply src/source/src/connector_source.rs:221]: Or is it a re-export?

Comment (xxchan) [reply src/source/src/connector_source.rs:221]: [Note that configuration changes will not apply for code that has already been compiled and cached](https://github.com/rust-lang/rust-clippy#:~:text=Note%20that%20configuration%20changes%20will%20not%20apply%20for%20code%20that%20has%20already%20been%20compiled%20and%20cached)

Comment (wangrunji0408) [reply src/source/src/connector_source.rs:221]: No? Still have not managed to make it work.

Comment (xxchan) [reply src/source/src/connector_source.rs:221]: https://github.com/rust-lang/rust-clippy/issues/9558 Reported an issue to clippy 😇

**Review by skyzh — APPROVED** [review]

File: src/connector/src/source/kinesis/source/reader.rs [hunk 4]
```diff
@@ -22,32 +21,22 @@ use aws_sdk_kinesis::model::ShardIteratorType;
...
-}
-
-impl Drop for KinesisMultiSplitReader {
-    fn drop(&mut self) {
-        if let Some(handler) = self.consumer_handler.as_mut() {
-            handler.abort();
-        }
-    }
+    readers: Vec<KinesisSplitReader>,
 }
```
Comment (tabVersion) [inline src/connector/src/source/kinesis/source/reader.rs:40]: please remove the struct. It was for consuming multiple splits and now we do it in the source layer

**Review by tabVersion — APPROVED** [review]
LGTM!

File: src/connector/src/source/kinesis/source/reader.rs (hunk 4 above)
Comment (wangrunji0408) [reply src/connector/src/source/kinesis/source/reader.rs]: done in 6fd08e3. PTAL!

**Review by BugenZhao — APPROVED** [review]
L🚀TM!

**Pull request merged**